db/hints/manager.hh

```diff
@@ -317,6 +317,15 @@ private:
     bool draining_all() noexcept {
         return _state.contains(state::draining_all);
     }
+
+    /// Renames host directories named after IPs to host IDs.
+    ///
+    /// In the past, hosts were identified by their IPs. Now we use host IDs for that purpose,
+    /// but we want to ensure that old hints don't get lost if possible. This function serves
+    /// this purpose. It's only necessary when upgrading Scylla.
+    ///
+    /// This function should ONLY be called by `manager::start()`.
+    future<> migrate_ip_directories();
```

Comment:
What about failed upgrades?

Scenario:
1. upgrade a node
2. node migrates its hint directories
3. cold feet, downgrade
4. old version confused

Better to defer until the entire cluster is upgraded using a feature. This means we need to be able to replay using either format. The alternative is to give up on this, and just make sure older versions know to ignore the new format.

Reply:
@mykaul a node could be confused because it comes across a directory whose name doesn't represent a valid IP address, in which case there will be problems in the current code. That situation can happen when some of the nodes in the cluster get upgraded and start writing hints on disk, naming the directories using host IDs, but at some point the upgrade of another node fails. If I understand the documentation correctly, the whole cluster will undergo downgrade in that scenario. The created hint directories will remain, though.

I also think that out of those two approaches, ignoring those directories would be better. It should require far fewer changes and perhaps prevent introducing new bugs (compared to the former idea). However, if we cannot assume that the version the cluster is being upgraded from is the latest minor of the previous major, then (if I'm not mistaken) we can't simply add code responsible for ignoring those directories. If that's the case, a feature flag is the only solution I can see at the moment (even specifying in the documentation that hinted handoff should be disabled before the whole cluster has been upgraded wouldn't help because the hint managers for MVs will be created anyway).

However, if we _can_ add new code to the previous versions and assume the user uses it, then I think the easiest solution would be to do the exact same thing as in this PR – we try to map directories from host IDs to IPs if Hinted Handoff is still based on IPs, or the other way round if it's based on host IDs. That would boil down to cherry picking one commit (perhaps modified a bit if parameterising the code by template-like stuff makes it too complex).

Reply:
Requiring a minor upgrade before the major upgrade is a no-go IMO.

The new version is required to support ip-address based view hints, right?
So it needs this code path in any case and a cluster feature to enable host_id-based hint directories makes the most sense.

Reply:
I'm somewhat surprised how current code doesn't ignore directories whose name is not a valid IP address.

Reply:
Please let's not call this pull request to death because of the potential upgrade failure. If the ugprade fails, we lose hints. What is unacceptable with this? Please keep in mind hints are off by default now.

Reply:
@avikivity

Reply:
Possible solutions:
#### 1. Use a feature flag ####
Continue storing hints naming directories after IPs and only after the whole cluster has been upgraded, switch to naming them after host IDs. The biggest advantage of this approach is that no hints should be lost in the process. However, a price comes with it: between the start and end of an upgrade, we will have to store hints in directories named after IPs. When the upgrade finishes, we should rename them to host IDs. This part is tricky because new hints might still be generated, so hinted handoff would have to be turned off for some time to be able to move those directories. Also, it would make the code more complex because we'd have to handle both IPs and host IDs; and as of now, I can't give an accurate estimate of how much would have to be modified.

#### 2. Stop storing hints between the start and the end of an upgrade ####
Alternatively, when upgrading a node, we could disable hinted handoff when an upgrade starts and only when the whole cluster finishes the upgrade would we rename the directories and enable it back. This way all of the logic would correspond to using host IDs and the changes should be minimal. However, since hints would not be stored during an upgrade and that applies to MVs as well of course.

I discussed it with my team and we opt for the latter approach.

Reply:
@avikivity this comment has been left unaddressed for two weeks because it's not clear how to address it. 

In my view it doesn't need addressing in this series at all. This series started when the upgrade path to 6.0 was non-existent. So the upgrade problem for topology was solved while this series was being developed. Therefore I believe this series can follow hte same pattern and provide an upgrade path in a follow up.

But most importantly, I don't see the issue. Cold feet? Downgrade? Old version doesn't start because of the unknown files? Delete the hint files & restart. 

Keep in mind that hints are off by default. The chance of stepping into the scenario you describe is very low.

Reply:
@avikivity all in all with @dawmd suggestions and the suggestion to backpatch 5.x to ignore unknown files (it probably already does, so perhaps no back patching is needed), we have 4 options to choose from.

Please be so kind and choose a solution so that this pull request can get unblocked.

Reply:
Hints are required for materialized views to work correctly, and hints are written exactly during upgrades.

I think this merits a cluster feature.